sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/AppendFilesToTables.java

```diff
@@ -211,5 +213,31 @@ private ManifestWriter<DataFile> createManifestWriter(
                   tableLocation, manifestFilePrefix, uuid, spec.specId()));
       return ManifestFiles.write(spec, io.newOutputFile(location));
     }
+
+    // If bundle fails following a successful commit and gets retried, it may attempt to re-commit
+    // the same data.
+    // To mitigate, we check the files in this bundle and remove anything that was already
+    // committed in the last successful snapshot.
+    //
+    // TODO(ahmedabu98): This does not cover concurrent writes from other pipelines, where the
+    //  "last successful snapshot" might reflect commits from other sources. Ideally, we would make
+    //  this stateful, but that is update incompatible.
+    // TODO(ahmedabu98): add load test pipelines with intentional periodic crashing
+    private Iterable<FileWriteResult> removeAlreadyCommittedFiles(
+        Table table, Iterable<FileWriteResult> fileWriteResults) {
+      if (table.currentSnapshot() == null) {
+        return fileWriteResults;
+      }
+
+      List<String> committedFiles =
+          Streams.stream(table.currentSnapshot().addedDataFiles(table.io()))
```

Comment:
How many data files could you have per snapshot? Is there any chance this list gets large enough that it causes problems/is there any way to do filtering at the source instead of later in this process?

Reply:
Thanks for bringing this up, I later realized that we have an upstream GBK which means we'll have the same batch of files for each retry. We can just check for a single overlapping file instead of doing a full search

Reply:
Noting that this does still try to collect all the data files in the current snapshot into an in memory set - I think this is ok, though, it seems unlikely that this would fall over, and this does seem like a normal thing to do with iceberg.

Reply:
> it seems unlikely that this would fall over

I think so too, but updated to manually iterate over the files for good practice anyways